rfd/0191-workload-id-config-ux.md

```diff
@@ -0,0 +1,1454 @@
+  --destination /opt/workload-id
+
+ $ tbot start workload-identity \
+  --proxy-server example.teleport.sh:443 \
+  --join-token my-join-token \
+  --workload-identity-labels *:* \
+  --destination /opt/workload-id
+```
+
+### Configuration Tooling 
+
+The WorkloadIdentity resource will support configuration via four methods:
+
+- The `tctl create`/`tctl update` commands with a YAML representation 
+- The Teleport Terraform Provider
+- The Teleport Kubernetes Operator as Kubernetes CRDs
+- The Teleport gRPC API
+
+### Diagnostics Tooling
+
+The flexibility of the templating, rules and attributes system allows 
+potentially complex configurations to be created. As such, we must provide
+tooling to assist operators in testing and debugging WorkloadIdentity
+configurations.
+
+The `tctl` CLI will be extended with a new command
+`tctl workload-identity test` that will locally evaluate a WorkloadIdentity,
+or set of WorkloadIdentities, against a given set of attributes and return a
+description of the workload identity credentials that would have been issued.
+
+Operators will be able to specify a WorkloadIdentity from a local file which
+may not yet exist in the Teleport cluster, or specify an existing
+WorkloadIdentity by name.
+
+Operators will provide the attributes as a file in YAML or JSON format. This
+will be in the same structure as the attributes included in the Teleport
+audit log.
+
+For example:
+
+```shell
+$ tctl workload-identity test --workload-identity-file ./my-workload-identity.yaml --attributes-file ./attributes.yaml
+Evaluating 3 WorkloadIdentity resources against the following attributes:
+--snip--
+
+1 WorkloadIdentity resources matched the given attributes. The following
+credentials would have been issued:
+
+- workload_identity_name: gitlab-production
+  spiffe:
+    id: spiffe://example.teleport.sh/gitlab/my-org/my-project/production
+    jwt:
+      sub: spiffe://example.teleport.sh/gitlab/my-org/my-project/production
+      iss: https://example.teleport.sh
+      -- snipped for conciseness: all JWT claims would be included --
+    x509:
+      subject:
+        common_name: spiffe://example.teleport.sh/gitlab/my-org/my-project/production
+        organization: my-org
+        -- snipped for conciseness: all Subject DN fields would be included --
+      sans:
+      - dns: my-project.example.com
+      - ip: 10.0.0.1
+      - uri: spiffe://example.teleport.sh/gitlab/my-org/my-project/production 
+2 WorkloadIdentity resources did not match the given attributes:
+
+- workload_identity_name: gitlab-staging
+  reason: Rule evaluation failed. Allow rule expression `join.gitlab.environment == "production"` returned false.
+- workload_identity_name: github-production
+  reason: The `join.github.environment` attribute included in `spec.spiffe.id` template do not exist in the attribute set.
+```
+
+It stands to reason that this functionality could eventually be integrated into 
+the Teleport Web UI.
```

Comment:
I think this is critical to make sure happens, rather than leaving it to "eventually." What are the resource requirements standing in the way? Would they be solved with the hires you want to make for this team, or would they require borrowing a full-stack from another team? What would be a realistic timeline?

Reply:
952b87991b